In [8]:
from statsmodels.stats.power import tt_ind_solve_power, zt_ind_solve_power
from statsmodels.stats.proportion import proportion_effectsize
from statsmodels.stats.meta_analysis import effectsize_smd
from typing import Union
import plotly.graph_objects as go
from scipy import stats
from math import asin
import numpy as np

Задача №1. Продакт на главной mail.ru решил протестировать в рекомендательной ленте контента вместо карточек со статьями видеоплеер с короткими видео. Нынешний таймспент на юзера в день в среднем равен 25 минут, а стандартная ошибка (SD) равна 15. Мы предполагаем, что в новой версии таймспент на юзера в день изменится на 10% от текущего значения. Средний трафик  20000 человек в день. Посчитайте сколько дней необходимо держать эксперимент при alpha = 5% и beta = 20%.

In [9]:
nobs = 20_000
mean_1, mean_2 = 25, 27.5
std_1, std_2 = 15, 15 * 1.15
nobs_1, nobs_2 =10_000, 10_000
effect_size = effectsize_smd(mean_1, std_1, nobs_1, mean_2, std_2, nobs_2 )[0]
effect_size

-0.15465700582340247

In [10]:
def calc_sample_size_continuous(effect_size: float,
                                alpha: float = .05,
                                beta: float = .2,
                                ratio: Union[float, int] = 1):
    
    n = tt_ind_solve_power(effect_size=effect_size,
                           alpha=alpha,
                           power=(1 - beta),
                           ratio=ratio,
                  )
    return int(n * 2)

In [11]:
sample = calc_sample_size_continuous(effect_size=effect_size, alpha=.05, beta=.2)
sample

1314

Правда немного не понял про дни, получается тест должен длится 1314 дней? Или я что то не то сделал?

Задача №2. Наша продуктовая команда в ecommerce магазине планирует запустить тест, направленный на ускорение загрузки сайта. Одна из основных метрик bounce rate в GA = 40%. Мы предполагаем, что при оптимизации сайта изменится минимум на 20% от текущего показателя.Средний трафик  4000 человек в день. Посчитайте сколько нам нужно дней держать эксперимент при alpha = 5% и beta = 20%


In [12]:
nobs = 4_000
conv_1, conv_2 = 800, 1200
nobs_1, nobs_2 = 2_000, 2_000
prob_1, prob_2 = conv_1/nobs_1, conv_2/nobs_2
effect_size = proportion_effectsize(prob_1, prob_2)
print(prob_1, prob_2)
effect_size



0.4 0.6


-0.40271584158066154

In [13]:
def calc_sample_size_proportion(effect_size: float,
                                alpha: float = .05,
                                beta: float = .2,
                                ratio: Union[float, int] = 1):
    
    n = zt_ind_solve_power(effect_size=effect_size,
                           alpha=alpha,
                           power=(1 - beta),
                           ratio=ratio,
                  )
    return int(n * 2)

In [14]:
sample = calc_sample_size_proportion(effect_size=effect_size, alpha = .05, beta = .2)
sample

193